corehq/util/metrics/prometheus.py

```diff
@@ -0,0 +1,46 @@
+from typing import Iterable
+
+import settings
+from corehq.util.metrics.metrics import (
+    HqCounter,
+    HqGauge,
+    HqMetrics,
+    MetricBase,
+)
+from prometheus_client import Counter as PCounter
+from prometheus_client import Gauge as PGauge
+
+
+class PrometheusMetricBase(MetricBase):
+    _metric_class = None
+
+    def __init__(self, name: str, documentation: str, tag_names: Iterable=tuple(), tag_values=None, delegate=None):
+        name = name.replace('.', '_')
+        super().__init__(name, documentation, tag_names, tag_values)
+        self._delegate = delegate or self._metric_class(name, documentation, tag_names, labelvalues=tag_values)
+
+    def _get_tagged_instance(self, tag_values):
+        delegate = self._delegate.labels(dict(zip(self.tag_names, tag_values)))
+        return self.__class__(self.name, self.documentation, self.tag_names, self.tag_values, delegate)
+
+
+class Counter(PrometheusMetricBase, HqCounter):
```

Comment:
Yes, the inheritance isn't great here. The alternative is a bit more duplicate code but not a lot. I'll see what I can do to make it better if we stay on this course.

If we want to keep it functional then we would need to have a centralized metrics singleton that would always return the same metric class when called (in the case of Pometheus). It would also mean that every call to the metric would need to pass in all required args (which is the case now anyway):
```python
metrics_histogram(value, tags=..., buckets=..., unit=..., dd_bucket_tag=...)
```

Reply:
I've resolved this by removing the need to create new copies of the class when applying tags:
```diff
- metric.tag(group='a').inc()
+ metric.inc(group='a')
```